stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -709,24 +693,19 @@ julia> dot([im; im], [1; 1])
 0 - 2im
 ```
 """
-function dot(x::AbstractVector, y::AbstractVector)
-    if length(LinearIndices(x)) != length(LinearIndices(y))
-        throw(DimensionMismatch("dot product arguments have unequal lengths $(length(LinearIndices(x))) and $(length(LinearIndices(y)))"))
+function dot(x::AbstractArray, y::AbstractArray)
+    lx = _length(x)
+    if lx != _length(y)
+        throw(DimensionMismatch("first array has length $(lx) which does not match the length of the second, $(_length(y))."))
     end
-    ix = iterate(x)
-    if ix === nothing
-        # we only need to check the first vector, since equal lengths have been asserted
+    if lx == 0
         return dot(zero(eltype(x)), zero(eltype(y)))
     end
-    iy = iterate(y)
-    s = dot(ix[1], iy[1])
-    ix, iy = iterate(x, ix[2]), iterate(y, iy[2])
-    while ix != nothing
-        s += dot(ix[1], iy[1])
-        ix = iterate(x, ix[2])
-        iy = iterate(y, iy[2])
+    s = zero(dot(first(x), first(y)))
```

Comment:
Seems the only potential downside to this implementation is that it retrieves the first element of each of `x` and `y` twice, rather than once with the explicit `iterate` implementations?